### PR TITLE
Change global shortcuts for linux

### PR DESCRIPTION
**Pull request by fcastilloec**
This is a workaround to fix issue #5 
There seems to be a bug with electron and how it handles shortcuts that are already registered by the OS. Electron won't register any shortcuts that are already being used by another app or the OS, so normal media shortcuts won't work.
Changing the shortcuts to require `CTRL+(media key)` fixes the issue and works without a problem.

This seems to be a problem in other OS too according to some of the issues posted on the electron project but I wasn't able to confirm this myself.
You can see that other electron apps have the same problem here: [nukeop/nuclear#18](https://github.com/nukeop/nuclear/issues/18) and [twolfson/google-music-electron#45](https://github.com/twolfson/google-music-electron/issues/45)

### CODE REVIEW

**Comment by danielravina** [comment]
Looking at their mpris code: https://github.com/twolfson/google-music-electron/blob/master/lib/mpris.js It looks quite straight forward. I think we should still implement this if possible.

As you mentioned in your comment https://github.com/twolfson/google-music-electron/issues/45#issuecomment-318943475 users wouldn't know that there's a different shortcut and will think this is a bug. Perhaps we can still use the `globalShortcut` as a fallback if users don't have mpirs installd (using a similar try/catch method). WDYT?

**Comment by fcastilloec** [comment]
Sounds like a great idea! Feel free to disregard this PR and close

**Comment by danielravina** [comment]
You can use the same PR. I don't mind

**Comment by fcastilloec** [comment]
Based on some help provided by the electron people. I believe I have fixed this issue without having to use `mpris` at all. I'm just using `dbus` and it's working great with the native shortcuts.
I'll be testing it a little bit more tomorrow and then I'll probably post it.
Since I'm a newbie regarding electron and javascript in general, I'm double checking the code.

**Comment by fcastilloec** [comment]
@danielravina Please review/test the code and feel free to make any changes (especially on the comments I added).
If you get errors that `dbus` is using a different version, just run the rebuild script to fix it.
This code will fix all issues and it doesn't require `mpris` support.

File: debian/package.json [hunk 1]
```diff
@@ -3,12 +3,14 @@
...
   "main": "index.js",
+  "homepage": "http://headsetapp.co",
   "scripts": {
     "start": "NODE_ENV=development NODE_PATH=. nf start",
     "electron": "./node_modules/.bin/electron .",
     "player": "./bin/player-window",
     "production": "NODE_ENV=production electron .",
-    "build": "NODE_ENV=production NEW_VERSION=$npm_package_version ./bin/build"
+    "build": "NODE_ENV=production NEW_VERSION=$npm_package_version ./bin/build",
+    "rebuild": "electron-rebuild"
```
Comment (danielravina) [inline debian/package.json]: what is this for?

File: debian/index.js [hunk 2]
```diff
@@ -76,29 +77,24 @@ const start = () => {
         exec('kill -9 $(pgrep Headset) &> /dev/null')
       }
     })
+    
+    // Code based on MarshallOfSound Google-Play-Music-Desktop-Player-UNOFFICIAL-
+    // file src/main/features/linux/mediaKeysDBus.js
```
Comment (danielravina) [inline debian/index.js]: it's good that you mention that. It' better to just provide the file to the link on github

**Comment by danielravina** [comment]
Thanks. I will try it out. I might clean some of the formattings in the code. other than that. looks good 👍

File: debian/index.js (hunk 2 above)
Comment (danielravina) [inline debian/index.js]: same for the second comment.

File: debian/index.js [hunk 3]
```diff
@@ -140,3 +136,29 @@ ipcMain.on('player2Win', (e, args) => {
...
+          case 'Play': win.webContents.executeJavaScript(`
+                         window.electronConnector.emit('play-pause')
+                       `); return;
+          default: return;
+        }
+      });
+      iface.GrabMediaPlayerKeys(0, `org.${desktopEnv}.SettingsDaemon.MediaKeys`);
+    }
+  });
+};
```
Comment (danielravina) [inline debian/index.js]: I would extract this function to a module and format it like so:
```js
// registerBindings.js
module.exports = registerBindings(desktopEnv, bus) {
  const arg1 = `org.${desktopEnv}.SettingsDaemon`
  const arg2 = `/org/${desktopEnv}/SettingsDaemon/MediaKeys`
  const arg3 = `org.${desktopEnv}.SettingsDaemon.MediaKeys`

  bus.getInterface(arg1, arg2, arg3, (err, interface) => {
    if (err) return
    interface.on('MediaPlayerKeyPressed', (n, keyName) => {
      switch (keyName) {
        case 'Next': 
          win.webContents.executeJavaScript(`
            window.electronConnector.emit('play-next')
          `); 
          break;
        case 'Previous': 
          win.webContents.executeJavaScript(`
            window.electronConnector.emit('play-previous')
          `);
          break;
        case 'Play': 
          win.webContents.executeJavaScript(`
            window.electronConnector.emit('play-pause')
          `);
          break;
        default:
          break;
      }
    });
    
    interface.GrabMediaPlayerKeys(0, arg3);
  });
};
```

File: debian/package.json (hunk 1 above)
Comment (fcastilloec) [inline debian/package.json]: It's just a shortcut for rebuilding `dbus` module. This has to happen only once after running `npm install`. Feel free to delete the line if you want to. I won't make a difference

File: debian/index.js (hunk 3 above)
Comment (danielravina) [inline debian/index.js]: I don't really know what arg1,arg2,arg3 mean. maybe you can give them a better name

File: debian/index.js (hunk 2 above)
Comment (fcastilloec) [inline debian/index.js]: Ok, like I said, I wasn't sure on the proper way to give credit for it

File: debian/package.json (hunk 1 above)
Comment (danielravina) [inline debian/package.json]: sure. I missed your previous comment

Comment (danielravina) [inline debian/package.json]: oh Bytheway, can you use `yarn add` instead of `npm install`? Thanks :)

File: debian/index.js (hunk 3 above)
Comment (fcastilloec) [inline debian/index.js]: According to the API for the `dbus` module:

```
arg1 = serviceName - The well-known name of the service that owns the object.
arg2 = objectPath - The path of the object.
arg3 = interfaceName - Which of the object's interfaces to retrieve.
```
Hope that helps with the naming.

File: debian/package.json (hunk 1 above)
Comment (fcastilloec) [inline debian/package.json]: I've never heard of yarn until now. Like I mentioned, I'm a newbie at electron and javascript. I'll install it on my system and see how it goes/works

File: debian/index.js (hunk 3 above)
Comment (danielravina) [inline debian/index.js]: cool! let's use these names

File: debian/index.js [hunk 4]
```diff
@@ -76,29 +77,24 @@ const start = () => {
...
+    // file src/main/features/linux/mediaKeysDBus.js
+    // Other desktop environments might have their own dbus which can be added
+    try {
+      const bus = DBus.getBus('session');
+
+      registerBindings('gnome', bus);
+      registerBindings('mate', bus);
+      
+    } catch (err) {
+
```
Comment (m1guelpf) [inline debian/index.js]: Ehy are we silently failing here?

**Comment by fcastilloec** [comment]
Hey guys, I've pushed the changes you mentioned in the comments with some extra modifications to make it work properly.
I was getting an error from the `dbus` module:  

> Error: No introspectable
>     at [path]/headset-electron/debian/node_modules/dbus/lib/bus.js:129:15

This error is only shown when starting the app and it doesn't interfere with the normal operation of the module or headset. The MediaKeys keep working. Deleting line 14 on `registerBindings.js` keeps the console clean.
This error was also being thrown on the other apps, and that's why they weren't being sent to the console. I tried googling everywhere about it, but there's nothing consistent regarding what's going on. I also tried using an older version of  `dbus`, v0.2.23, but the same problem happened.

**Comment by danielravina** [comment]
I merged that in. I think if the error is not critical, we can silence it.

**Comment by fcastilloec** [comment]
Great! I'll create a new pull request with some extra fixes for Linux, so the deb packages have fewer dependencies. Also, I'll submit the Travis setup on that new pull request

**Pull request merged**